README.md
# TWICE Bootanimation
[Demo](https://t.me/MrMiss_Repo/45)

## What is this
Magisk module to change your android bootanimation

### Credits
- topjohnwu for magisk
- Zackptg5 for MMT Extended


changelog.md
### v3.3
* Add support for magisk 24 module update method.

### v3.2
* Update bootanimation upstream link.

### v3.1
* Disable installing debug log.

### v3.0
* Add user config selection on install screen.
    * Add volume key selector.
* Bootanimation is now downloaded on module installation (avoiding downloading the whole bootanimation file).
    * Add curl lib to download bootanimation.

### v2.0
* Magisk flashable zip.
* Resized down bootanimation file.

### v1.0
* Initial release.

customize.sh
##########################################################################################
#
# MMT Extended Config Script
#
##########################################################################################

##########################################################################################
# Config Flags
##########################################################################################

# Uncomment and change 'MINAPI' and 'MAXAPI' to the minimum and maximum android version for your mod
# Uncomment DYNLIB if you want libs installed to vendor for oreo+ and system for anything older
# Uncomment DEBUG if you want full debug logs (saved to /sdcard)
MINAPI=26
#MAXAPI=
#DYNLIB=true
#DEBUG=true

##########################################################################################
# Replace list
##########################################################################################

# List all directories you want to directly replace in the system
# Check the documentations for more info why you would need this

# Construct your list in the following format
# This is an example
REPLACE_EXAMPLE="
/system/app/Youtube
/system/priv-app/SystemUI
/system/priv-app/Settings
/system/framework
"

# Construct your own list here
REPLACE="
"

##########################################################################################
# Permissions
##########################################################################################

set_permissions() {
  : # Remove this if adding to this function

  # Note that all files/folders in magisk module directory have the $MODPATH prefix - keep this prefix on all of your files/folders
  # Some examples:
  
  # For directories (includes files in them):
  # set_perm_recursive  <dirname>                <owner> <group> <dirpermission> <filepermission> <contexts> (default: u:object_r:system_file:s0)
  
  # set_perm_recursive $MODPATH/system/lib 0 0 0755 0644
  # set_perm_recursive $MODPATH/system/vendor/lib/soundfx 0 0 0755 0644

  # For files (not in directories taken care of above)
  # set_perm  <filename>                         <owner> <group> <permission> <contexts> (default: u:object_r:system_file:s0)
  
  # set_perm $MODPATH/system/lib/libart.so 0 0 0644
  # set_perm /data/local/tmp/file.txt 0 0 644
}

##########################################################################################
# MMT Extended Logic - Don't modify anything after this
##########################################################################################

SKIPUNZIP=1
unzip -qjo "$ZIPFILE" 'common/functions.sh' -d $TMPDIR >&2
. $TMPDIR/functions.sh

# Custom Variables for Install AND Uninstall - Keep everything within this function - runs before uninstall/install
# This is Script is kanged from fillipo, the creator of excellent module Oxify.
custom() {


 # Build Prop Reader:

   if [ -f $VEN/build.prop ]; then BUILDS="/system/build.prop $VEN/build.prop"; else BUILDS="/system/build.prop"; fi
  PIXEL=$(grep -E "ro.product.manufacturer=Google|ro.product.vendor.brand=Google" "$BUILDS")
  SAMSUNG=$(grep -E "ro.product.manufacturer=Samsung|ro.product-vendor.brand=Samsung" "$BUILDS")
  OPPO=$(grep -E "ro.product.manufacturer=OPPO|ro.product.brand=OPPO" "$BUILDS")
  SONY=$(grep -E "ro.product.vendor.manufacturer=Sony|ro.product.vendor.brand=Sony" "$BUILDS")

# Easy directory values:

 S=/system
 SP=/system/product
 P=/product
 F=/fonts
 MA=/media/audio
 A=/app
 L=/priv-app
 OP=oppo_product
 NY=oem
}

common/install.sh
#Choose user bootanimation
ui_print "   ***********************************"
ui_print "        Choose your Bootanimation     "
ui_print ""
ui_print "     Vol [+] = dark, Vol [-] = light  "
ui_print "   ***********************************"
if $VKSEL; then
  theme="-dark"
  ui_print ""
  ui_print "Dark bootanimation selected"
  ui_print ""
else
  theme=""
  ui_print ""
  ui_print "light bootanimation selected"
  ui_print ""
fi

ui_print "   ****************************************************"
ui_print "                  Choose The resolution                "
ui_print "    better to choose the same resolution as your phone "
ui_print ""
ui_print "      Vol [+] = 1080p, Vol [-] = next(720p / 1440p)    "
ui_print "   ****************************************************"
if $VKSEL; then
  reso="1080"
  ui_print ""
  ui_print "    1080 selected"
  ui_print ""
else
  ui_print ""
  ui_print "               Vol [+] = 720p, Vol [-] = 1440p "
  ui_print "   ****************************************************"
  if $VKSEL; then
    reso="720"
    ui_print ""
    ui_print "    720 selected"
    ui_print ""
  else
    reso="1440"
    ui_print ""
    ui_print "    1440 selected"
    ui_print ""
  fi
fi

chmod 0755 $MODPATH/curl

#Download Bootaniamation
src_link="https://raw.githubusercontent.com/MrMissx/bootanimation/master/bootanimation"
DLINK="$src_link""$theme"_"$reso".zip
config="bootanimation${theme}_${reso}.zip"
ui_print "- Downloading ${config}"
# ui_print $DLINK
($MODPATH/curl -f -k -o $MODPATH/bootanimation.zip $DLINK) || abort "   Download failed!"
ui_print "- Done..."
ui_print ""

#Bootanimation installation:
ui_print "Installing Bootanimation..."
  
#Mount functions, to monut /oem in sony devices. Thanks a ton to: (https://github.com/ironydelerium)

is_mounted_rw() {
  cat /proc/mounts | grep -q " `readlink -f $1` " | grep -q " rw," 2>/dev/null
  return $?
}

mount_rw() {
  mount -o remount,rw $1
  DID_MOUNT_RW=$1
}

# is_mounted script thanks to (https://github.com/skittles9823)
if [ -d /oem/media ]; then
      is_mounted " /oem" || mount /oem
      is_mounted_rw " /oem" || mount_rw /oem
      is_mounted " /oem/media" || mount /oem/media
      is_mounted_rw " /oem/media" || mount_rw /oem/media
      mv /oem/media/bootanimation.zip /oem/media/bootanimation/bootanimation.zip
      cp_ch $MODPATH/bootanimation.zip /oem/media
fi

#Andoird 10 Multiple bootanimation support script:
for i in /system/media/*bootanimation*; do
    cp_ch $MODPATH/bootanimation.zip $MODPATH/$i
  if [ -f /system/media/bootanimation-dark.zip ]; then
    cp_ch $MODPATH/bootanimation.zip $MODPATH/system/media/bootanimation-dark.zip
  fi
done

for d in /system/product/media/*bootanimation*; do
  cp_ch $MODPATH/bootanimation.zip $MODPATH/$d
  if [ -f /system/product/media/bootanimation-dark.zip ]; then
    cp_ch $MODPATH/bootanimation.zip $MODPATH/system/product/media/bootanimation-dark.zip
  fi
done

ui_print ""
ui_print "Bootanimation Installation complete"
ui_print "If you want to change beetween dark or light"
ui_print "You can just reflash this module :)"

common/functions.sh
##########################################################################################
#
# MMT Extended Utility Functions
#
##########################################################################################

abort() {
  ui_print "$1"
  rm -rf $MODPATH 2>/dev/null
  cleanup
  rm -rf $TMPDIR 2>/dev/null
  exit 1
}

cleanup() {
  rm -rf $MODPATH/common 2>/dev/null
  ui_print " "
  ui_print "**************************************"
  ui_print "*   MMT Extended by Zackptg5 @ XDA   *"
  ui_print "**************************************"
  ui_print " "
  $DEBUG && debug_log
}

debug_log() {
  set +x
  echo -e "***---Device Info---***" > $LOGFILE-tmp.log
  echo -e "\n---Props---\n" >> $LOGFILE-tmp.log
  getprop >> $LOGFILE-tmp.log
  echo -e "\n\n***---Magisk Info---***" >> $LOGFILE-tmp.log
  echo -e "\n---Magisk Version---\n\n$MAGISK_VER_CODE" >> $LOGFILE-tmp.log
  echo -e "\n---Installed Modules---\n" >> $LOGFILE-tmp.log
  ls $NVBASE/modules >> $LOGFILE-tmp.log
  echo -e "\n---Last Magisk Log---\n" >> $LOGFILE-tmp.log
  cat /cache/magisk.log >> $LOGFILE-tmp.log
  echo -e "\n\n***---MMT Extended Debug Info---***" >> $LOGFILE-tmp.log
  if [ -d "$MODPATH" ]; then
    echo -e "\n---Installed Files---\n" >> $LOGFILE-tmp.log
    grep "^+* cp_ch" $LOGFILE.log | sed 's/.* //g' >> $LOGFILE-tmp.log
    sed -i -e "\|$TMPDIR/|d" -e "\|$MODPATH|d" $LOGFILE-tmp.log
    find $MODPATH -type f >> $LOGFILE-tmp.log
    echo -e "\n---Installed Boot Scripts---\n" >> $LOGFILE-tmp.log
    grep "^+* install_script" $LOGFILE.log | sed -e 's/.* //g' -e 's/^-.* //g' >> $LOGFILE-tmp.log
    echo -e "\n---Installed Prop Files---\n" >> $LOGFILE-tmp.log
    grep "^+* prop_process" $LOGFILE.log | sed 's/.* //g' >> $LOGFILE-tmp.log
  fi
  echo -e "\n---Shell & MMT Extended Variables---\n" >> $LOGFILE-tmp.log
  (set) >> $LOGFILE-tmp.log
  echo -e "\n---(Un)Install Log---\n" >> $LOGFILE-tmp.log
  echo "$(cat $LOGFILE.log)" >> $LOGFILE-tmp.log
  mv -f $LOGFILE-tmp.log $LOGFILE.log
}

device_check() {
  local PROP=$(echo "$1" | tr '[:upper:]' '[:lower:]')
  for i in /system /vendor /odm /product; do
    if [ -f $i/build.prop ]; then
      for j in "ro.product.device" "ro.build.product" "ro.product.vendor.device" "ro.vendor.product.device"; do
        [ "$(sed -n "s/^$j=//p" $i/build.prop 2>/dev/null | head -n 1 | tr '[:upper:]' '[:lower:]')" == "$PROP" ] && return 0
      done
    fi
  done
  return 1
}

run_addons() {
  local OPT=`getopt -o mpi -- "$@"` NAME PNAME
  eval set -- "$OPT"
  while true; do
    case "$1" in
      -m) NAME=main; shift;;
      -p) NAME=preinstall; PNAME="Preinstall"; shift;;
      -i) NAME=install; PNAME="Install"; shift;;
      --) shift; break;;
    esac
  done
  if [ "$(ls -A $MODPATH/common/addon/*/$NAME.sh 2>/dev/null)" ]; then
    [ -z $PNAME ] || { ui_print " "; ui_print "- Running $PNAME Addons -"; }
    for i in $MODPATH/common/addon/*/$NAME.sh; do
      ui_print "  Running $(echo $i | sed -r "s|$MODPATH/common/addon/(.*)/$NAME.sh|\1|")..."
      . $i
    done
    [ -z $PNAME ] || { ui_print " "; ui_print "- `echo $PNAME`ing (cont) -"; }
  fi
}

cp_ch() {
  local OPT=`getopt -o inr -- "$@"` BAK=true UBAK=true FOL=false
  eval set -- "$OPT"
  while true; do
    case "$1" in
      -n) UBAK=false; shift;;
      -r) FOL=true; shift;;
      --) shift; break;;
      *) abort "Invalid cp_ch argument $1! Aborting!";;
    esac
  done
  local SRC="$1" DEST="$2" OFILES="$1"
  $FOL && local OFILES=$(find $SRC -type f 2>/dev/null)
  [ -z $3 ] && PERM=0644 || PERM=$3
  case "$DEST" in
    $TMPDIR/*|$MODULEROOT/*|$NVBASE/modules/$MODID/*) BAK=false;;
  esac
  for OFILE in ${OFILES}; do
    if $FOL; then
      if [ "$(basename $SRC)" == "$(basename $DEST)" ]; then
        local FILE=$(echo $OFILE | sed "s|$SRC|$DEST|")
      else
        local FILE=$(echo $OFILE | sed "s|$SRC|$DEST/$(basename $SRC)|")
      fi
    else
      [ -d "$DEST" ] && local FILE="$DEST/$(basename $SRC)" || local FILE="$DEST"
    fi
    if $BAK && $UBAK; then
      [ ! "$(grep "$FILE$" $INFO 2>/dev/null)" ] && echo "$FILE" >> $INFO
      [ -f "$FILE" -a ! -f "$FILE~" ] && { mv -f $FILE $FILE~; echo "$FILE~" >> $INFO; }
    elif $BAK; then
      [ ! "$(grep "$FILE$" $INFO 2>/dev/null)" ] && echo "$FILE" >> $INFO
    fi
    install -D -m $PERM "$OFILE" "$FILE"
  done
}

install_script() {
  case "$1" in
    -l) shift; local INPATH=$NVBASE/service.d;;
    -p) shift; local INPATH=$NVBASE/post-fs-data.d;;
    *) local INPATH=$NVBASE/service.d;;
  esac
  [ "$(grep "#!/system/bin/sh" $1)" ] || sed -i "1i #!/system/bin/sh" $1
  local i; for i in "MODPATH" "LIBDIR" "MODID" "INFO" "MODDIR"; do
    case $i in
      "MODPATH") sed -i "1a $i=$NVBASE/modules/$MODID" $1;;
      "MODDIR") sed -i "1a $i=\${0%/*}" $1;;
      *) sed -i "1a $i=$(eval echo \$$i)" $1;;
    esac
  done
  [ "$1" == "$MODPATH/uninstall.sh" ] && return 0
  case $(basename $1) in
    post-fs-data.sh|service.sh) ;;
    *) cp_ch -n $1 $INPATH/$(basename $1) 0755;;
  esac
}

prop_process() {
  sed -i -e "/^#/d" -e "/^ *$/d" $1
  [ -f $MODPATH/system.prop ] || mktouch $MODPATH/system.prop
  while read LINE; do
    echo "$LINE" >> $MODPATH/system.prop
  done < $1
}

# Check for min/max api version
[ -z $MINAPI ] || { [ $API -lt $MINAPI ] && abort "! Your system API of $API is less than the minimum api of $MINAPI! Aborting!"; }
[ -z $MAXAPI ] || { [ $API -gt $MAXAPI ] && abort "! Your system API of $API is greater than the maximum api of $MAXAPI! Aborting!"; }

# Set variables
[ $API -lt 26 ] && DYNLIB=false
[ -z $DYNLIB ] && DYNLIB=false
[ -z $DEBUG ] && DEBUG=false
[ -e "$PERSISTDIR" ] && PERSISTMOD=$PERSISTDIR/magisk/$MODID
INFO=$NVBASE/modules/.$MODID-files
ORIGDIR="$MAGISKTMP/mirror"
if $DYNLIB; then
  LIBPATCH="\/vendor"
  LIBDIR=/system/vendor
else
  LIBPATCH="\/system"
  LIBDIR=/system
fi
if ! $BOOTMODE; then
  ui_print "- Only uninstall is supported in recovery"
  ui_print "  Uninstalling!"
  touch $MODPATH/remove
  [ -s $INFO ] && install_script $MODPATH/uninstall.sh || rm -f $INFO $MODPATH/uninstall.sh
  recovery_cleanup
  cleanup
  rm -rf $NVBASE/modules_update/$MODID $TMPDIR 2>/dev/null
  exit 0
fi

# Debug
if $DEBUG; then
  ui_print "- Debug mode"
  LOGFILE=/storage/emulated/0/Download/$MODID-debug
  ui_print "  Debug log will be written to: $LOGFILE.log"
  exec 2>$LOGFILE.log
  set -x
fi

# Extract files
ui_print "- Extracting module files"
unzip -o "$ZIPFILE" -x 'META-INF/*' 'common/functions.sh' -d $MODPATH >&2
[ -f "$MODPATH/common/addon.tar.xz" ] && tar -xf $MODPATH/common/addon.tar.xz -C $MODPATH/common 2>/dev/null

# Main addons
run_addons -m

# Remove files outside of module directory
ui_print "- Removing old files"

if [ -f $INFO ]; then
  while read LINE; do
    if [ "$(echo -n $LINE | tail -c 1)" == "~" ]; then
      continue
    elif [ -f "$LINE~" ]; then
      mv -f $LINE~ $LINE
    else
      rm -f $LINE
      while true; do
        LINE=$(dirname $LINE)
        [ "$(ls -A $LINE 2>/dev/null)" ] && break 1 || rm -rf $LINE
      done
    fi
  done < $INFO
  rm -f $INFO
fi

### Install
ui_print "- Installing"

run_addons -p
[ -f "$MODPATH/common/install.sh" ] && . $MODPATH/common/install.sh
run_addons -i

ui_print "   Installing for $ARCH SDK $API device..."
# Remove comments from files and place them, add blank line to end if not already present
for i in $(find $MODPATH -type f -name "*.sh" -o -name "*.prop" -o -name "*.rule"); do
  [ -f $i ] && { sed -i -e "/^#/d" -e "/^ *$/d" $i; [ "$(tail -1 $i)" ] && echo "" >> $i; } || continue
  case $i in
    "$MODPATH/service.sh") install_script -l $i;;
    "$MODPATH/post-fs-data.sh") install_script -p $i;;
    "$MODPATH/uninstall.sh") [ -s $INFO ] && install_script $MODPATH/uninstall.sh || rm -f $INFO $MODPATH/uninstall.sh;;
  esac
done

$IS64BIT || for i in $(find $MODPATH/system -type d -name "lib64"); do rm -rf $i 2>/dev/null; done  
[ -d "/system/priv-app" ] || mv -f $MODPATH/system/priv-app $MODPATH/system/app 2>/dev/null
[ -d "/system/xbin" ] || mv -f $MODPATH/system/xbin $MODPATH/system/bin 2>/dev/null
if $DYNLIB; then
  for FILE in $(find $MODPATH/system/lib* -type f 2>/dev/null | sed "s|$MODPATH/system/||"); do
    [ -s $MODPATH/system/$FILE ] || continue
    case $FILE in
      lib*/modules/*) continue;;
    esac
    mkdir -p $(dirname $MODPATH/system/vendor/$FILE)
    mv -f $MODPATH/system/$FILE $MODPATH/system/vendor/$FILE
    [ "$(ls -A `dirname $MODPATH/system/$FILE`)" ] || rm -rf `dirname $MODPATH/system/$FILE`
  done
  # Delete empty lib folders (busybox find doesn't have this capability)
  toybox find $MODPATH/system/lib* -type d -empty -delete >/dev/null 2>&1
fi

# Set permissions
ui_print " "
ui_print "- Setting Permissions"
set_perm_recursive $MODPATH 0 0 0755 0644
if [ -d $MODPATH/system/vendor ]; then
  set_perm_recursive $MODPATH/system/vendor 0 0 0755 0644 u:object_r:vendor_file:s0
  [ -d $MODPATH/system/vendor/app ] && set_perm_recursive $MODPATH/system/vendor/app 0 0 0755 0644 u:object_r:vendor_app_file:s0
  [ -d $MODPATH/system/vendor/etc ] && set_perm_recursive $MODPATH/system/vendor/etc 0 0 0755 0644 u:object_r:vendor_configs_file:s0
  [ -d $MODPATH/system/vendor/overlay ] && set_perm_recursive $MODPATH/system/vendor/overlay 0 0 0755 0644 u:object_r:vendor_overlay_file:s0
  for FILE in $(find $MODPATH/system/vendor -type f -name *".apk"); do
    [ -f $FILE ] && chcon u:object_r:vendor_app_file:s0 $FILE
  done
fi
set_permissions

# Complete install
cleanup
